cairo-libs/packages/protobuf/src/tests/oneof.cairo

```diff
@@ -0,0 +1,127 @@
+use protobuf::types::message::{
+    ProtoMessage, ProtoOneof, DecodeContext, EncodeContext, EncodeContextTrait, DecodeContextTrait,
+    ProtoCodecImpl,
+};
+use protobuf::types::tag::{WireType, ProtobufTag};
+use protobuf::primitives::numeric::U64AsProtoMessage;
+use protobuf::primitives::array::ByteArrayAsProtoMessage;
+
+#[derive(Drop, Debug, Default, PartialEq)]
+pub struct MessageWithOneof {
+    oneof: Oneof,
+}
+
+pub impl MessageWithOneofAsProtoMessage of ProtoMessage<MessageWithOneof> {
+    fn encode_raw(self: @MessageWithOneof, ref context: EncodeContext) {
+        context.encode_oneof(self.oneof);
+    }
+
+    fn decode_raw(ref context: DecodeContext) -> Option<MessageWithOneof> {
+        let oneof = context.decode_oneof()?;
+        Option::Some(MessageWithOneof { oneof })
+    }
+
+    fn wire_type() -> WireType {
+        WireType::LengthDelimited
+    }
+}
+
+#[derive(Drop, Debug, Default, PartialEq)]
+pub enum Oneof {
+    #[default]
+    Int: u64,
+    String: ByteArray,
+    Enum: Enum,
+}
+
+pub impl OneofAsProtoOneof of ProtoOneof<Oneof> {
+    fn encode_raw(self: @Oneof, ref context: EncodeContext) -> ProtobufTag {
+        match self {
+            Oneof::Int(n) => {
+                n.encode_raw(ref context);
+                let wire_type = ProtoMessage::<u64>::wire_type();
+                ProtobufTag { field_number: 1, wire_type }
+            },
+            Oneof::String(str) => {
+                str.encode_raw(ref context);
+                let wire_type = ProtoMessage::<ByteArray>::wire_type();
+                ProtobufTag { field_number: 2, wire_type }
+            },
+            Oneof::Enum(e) => {
+                let n: u32 = e.into();
+                n.encode_raw(ref context);
+                let wire_type = ProtoMessage::<u32>::wire_type();
+                ProtobufTag { field_number: 3, wire_type }
+            },
+        }
+    }
```

Comment:
I am not sure why this looks like this. IMO it should look like this,

```cairo
        match self {
            Oneof::Int(n) => context.encode_field(1, n),
            Oneof::String(str) => context.encode_field(2, str),
            Oneof::Enum(e) => context.encode_field(3, e),
        }
```

Reply:
The thing is normal protobuf encoding works so that if a field is empty or unavailable ([i.e., it matches the default value](https://github.com/informalsystems/ibc-starknet/blob/17b0b7188ede8b0e1b44d4294f9ae86abc12267b/cairo-libs/packages/protobuf/src/types/message.cairo#L40)), it's ignored and doesn't appear in the result. However, for enums, the default value shouldn’t be ignored.

We might be able to simplify things based on your suggestion, but should the entire Protobuf implementation be refactored (out of this PR's scope) and decoupled from defaults (if possible). Might be better to deal with it as a lower-priority issue later.

Reply:
BTW, the alternative to the `Default` is opting for optional fields, as that's how `prost` handles it too.

Reply:
ah good catch ! 🙌🏼 

but I still think, this ruins the clean code. I see an immediate hack, which is, `context.encode_repeated_field(2, array![str])` which doesn't ignore the default value.

Maybe add one method `encode_field_forced()` to not ignore default values.

Reply:
It seems, the variant is broke that all the `ProtoMessage` can be used for `encode_field`. I can't use `Enum(e)` in `context.encode_enum(3, e)`. I have to call `context.encode_enum(3, e)`.

Reply:
I get what you're aiming for, but let's address this when it really matters for the project. I dropped my two cents in the issue you opened:  https://github.com/informalsystems/ibc-starknet/issues/323#issuecomment-2685606836